src/js/wrapper.ts

```diff
@@ -87,77 +90,40 @@ export const NATIVE: SentryNativeWrapper = {
       throw this._NativeClientError;
     }
 
-    const header = envelope[0];
+    const [EOL] = utf8ToBytes('\n');
 
-    if (NATIVE.platform === 'android') {
-      // Android
+    const [envelopeHeader, envelopeItems] = envelope;
 
-      const headerString = JSON.stringify(header);
+    const headerString = JSON.stringify(envelopeHeader);
+    const envelopeBytes: number[] = utf8ToBytes(headerString);
+    envelopeBytes.push(EOL);
 
-      let envelopeItemsBuilder = `${headerString}`;
+    let isFatalCrash: boolean = false;
+    for (const rawItem of envelopeItems) {
+      const [itemHeader, itemPayload] = this._processItem(rawItem);
 
-      for (const envelopeItems of envelope[1]) {
-
-        const event = this._getEvent(envelopeItems);
-        if (event != undefined) {
-
-          // @ts-ignore Android still uses the old message object, without this the serialization of events will break.
-          event.message = { message: event.message };
-
-          /*
-        We do this to avoid duplicate breadcrumbs on Android as sentry-android applies the breadcrumbs
-        from the native scope onto every envelope sent through it. This scope will contain the breadcrumbs
-        sent through the scope sync feature. This causes duplicate breadcrumbs.
-        We then remove the breadcrumbs in all cases but if it is handled == false,
-        this is a signal that the app would crash and android would lose the breadcrumbs by the time the app is restarted to read
-        the envelope.
-          */
-          if (event.exception?.values?.[0]?.mechanism?.handled != false && event.breadcrumbs) {
-            event.breadcrumbs = [];
-          }
-          envelopeItems[1] = event;
-        }
-
-        // Content type is not inside BaseEnvelopeItemHeaders.
-        (envelopeItems[0] as BaseEnvelopeItemHeaders).content_type = 'application/json';
-
-        const itemPayload = JSON.stringify(envelopeItems[1]);
-
-        let length = itemPayload.length;
-        try {
-          length = await RNSentry.getStringBytesLength(itemPayload);
-        } catch {
-          // The native call failed, we do nothing, we have payload.length as a fallback
-        }
-
-        (envelopeItems[0] as BaseEnvelopeItemHeaders).length = length;
-        const itemHeader = JSON.stringify(envelopeItems[0]);
-
-        envelopeItemsBuilder += `\n${itemHeader}\n${itemPayload}`;
+      let bytesPayload: number[] = [];
+      if (typeof itemPayload === 'string') {
+        bytesPayload = utf8ToBytes(itemPayload);
+      } else if (itemPayload instanceof Uint8Array) {
+        bytesPayload = [...itemPayload];
+      } else {
+        bytesPayload = utf8ToBytes(JSON.stringify(itemPayload));
+        isFatalCrash = 'level' in itemPayload && itemPayload.level === 'fatal';
```

Comment:
Let's create a new issue to improve this later, so we check the `mechanism.handled` instead, this logic could be extracted and also used for the Android breadcrumbs check, `wasExceptionHandled`.

Reply:
- https://github.com/getsentry/sentry-react-native/issues/2470